Extensions/Leaderboards/leaderboardstools.ts

```diff
@@ -120,61 +146,161 @@ namespace gdjs {
+                  "The leaderboard view couldn't be found. Doing nothing."
+                );
+                return;
+              }
+
+              _leaderboardViewIframe.style.opacity = '1';
+            }
+            _leaderboardViewIframeLoaded = true;
+            _leaderboardViewIframeLoading = false;
+
+            break;
         }
       };
 
-      export const displayLeaderboard = async function (
+      const onError = function (
         runtimeScene: gdjs.RuntimeScene,
-        leaderboardId: string
+        message: string
       ) {
-        const gameId = gdjs.projectData.properties.projectUuid;
-        const targetUrl = `https://liluo.io/games/${gameId}/leaderboard/${leaderboardId}`;
-        if (!(await checkLeaderboardAvailability(targetUrl))) {
-          logger.error('Leaderboard data could not be fetched, doing nothing');
-          return;
-        }
+        logger.error(message);
+        _leaderboardViewIframeErrored = true;
+        _leaderboardViewIframeLoading = false;
+        closeLeaderboardView(runtimeScene);
+      };
 
-        if (_leaderboardViewIframe) {
-          _leaderboardViewIframe.src = targetUrl;
-        } else {
-          const domElementContainer = runtimeScene
-            .getGame()
-            .getRenderer()
-            .getDomElementContainer();
-          if (!domElementContainer) {
-            logger.error(
-              "The div element covering the game couldn't be found, the leaderboard cannot be displayed."
+      const resetErrorTimeout = (runtimeScene: gdjs.RuntimeScene) => {
+        if (_errorTimeoutId) clearTimeout(_errorTimeoutId);
+        _errorTimeoutId = setTimeout(() => {
+          if (!_leaderboardViewIframeLoaded) {
+            onError(
+              runtimeScene,
+              'Leaderboard page did not send message in time. Closing leaderboard view.'
             );
-            return;
           }
-          const iframe = document.createElement('iframe');
-
-          iframe.src = targetUrl;
-          iframe.id = 'leaderboard-view';
-          iframe.style.position = 'absolute';
-          iframe.style.pointerEvents = 'all';
-          iframe.style.top = '0px';
-          iframe.style.height = '100%';
-          iframe.style.left = '0px';
-          iframe.style.width = '100%';
-          iframe.style.border = 'none';
-          _leaderboardViewIframe = iframe;
-          if (typeof window !== 'undefined') {
-            _leaderboardViewClosingCallback = (event: MessageEvent) => {
-              receiveMessage(runtimeScene, event);
-            };
-            (window as any).addEventListener(
-              'message',
-              _leaderboardViewClosingCallback,
-              true
+        }, 5000);
+      };
+
+      export const displayLeaderboard = function (
+        runtimeScene: gdjs.RuntimeScene,
+        leaderboardId: string,
+        displayLoader: boolean
+      ) {
+        _leaderboardViewIframeErrored = false;
```

Comment:
Last (?) edge case: if I call displayLeaderboard multiple times quickly with different leaderboard ids.
Like, `displayLeaderboard(.., 1)`, `displayLeaderboard(.., 2)` and finally `displayLeaderboard(.., 3)`.

There is small, but non zero, chance that 1 will take time and the response (of `checkLeaderboardAvailability`) will arrive after the response for 2 and 3.

So you will end up displaying the leaderboard 1 instead of the last one that was requested - which will be surprising and surely considered as a bug/strange behavior.

There is no simple way to avoid this - ideally we would have something to "cancel" promises that are "in flight" and stop execution of anything left.
As this is not easily doable in JS, you can still store the `_requestedLeaderboardId`. You set it directly at the beginning of the call - so we are assured this is up to date with the latest call.
When you received the response of checkLeaderboardAvailability, you check `if (_requestedLeaderboardId !== leaderboardId)`. If that's true, you are getting a response for something outdated. You simply bail out (and don't error - you just don't care about the response anymore).